src/vm_cpu.rs

```diff
@@ -211,31 +241,43 @@ impl VmPmuState {
 
     /// Updates internal state for PMU counters.
     /// This should be called following a successful SBI call to start counters.
-    pub fn update_started_counters(
-        &mut self,
-        counter_index: u64,
-        counter_mask: u64,
-        start_flags: PmuCounterStartFlags,
-        initial_value: u64,
-    ) {
+    pub fn update_started_counters(&mut self, counter_index: u64, counter_mask: u64) {
         let bitmask_iter = BitmaskIter::new(counter_mask);
         let mut updated = false;
-        self.counter_state
-            .iter_mut()
-            .skip(counter_index as usize)
-            .zip(bitmask_iter)
-            .for_each(|state_and_mask| {
-                let state = state_and_mask.0;
-                let selected = state_and_mask.1;
-                if let Configured(c) = state && selected == 1 {
-                    if start_flags.get_init_value() {
-                        c.value = initial_value;
+        let mut hcounteren_value = CSR.hcounteren.get();
+        // Unwrap ok: Success is guaranteed because because we must have verified it in an upper layer
+        let num_counters = pmu::get_num_counters().unwrap();
+        if counter_index < num_counters {
```

Comment:
Not sue what's the coding style in rust. But it is usually cleaner to return if counter_index >= num_counters so that core logic doesn't have to be indented for the conditional statement.